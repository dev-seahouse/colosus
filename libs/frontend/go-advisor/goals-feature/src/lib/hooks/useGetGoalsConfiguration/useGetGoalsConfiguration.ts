import type { QueryClient, UseQueryOptions } from '@tanstack/react-query';
import { useQuery } from '@tanstack/react-query';
import type { ConnectAdvisorGetGoalTypesResponseDto } from '@bambu/api-client';
import { ConnectAdvisorGoalTypesApi } from '@bambu/api-client';

async function fetchGoalsConfiguration() {
  const connectAdvisorGoalTypesApi = new ConnectAdvisorGoalTypesApi();
  const res = await connectAdvisorGoalTypesApi.getGoalTypes();
  return res.data;
}

export const getGoalConfigurationsQuery = () => ({
  queryKey: ['getGoalsConfiguration'],
  queryFn: fetchGoalsConfiguration,
  staleTime: 10 * 60 * 1000,
  gcTime: 15 * 60 * 1000,
  select: (data: ConnectAdvisorGetGoalTypesResponseDto) =>
    moveItemToEnd(data.goalTypes, 'Other'),
});

function moveItemToEnd<T extends { name: string }>(
  arr: readonly T[],
  itemName: string
) {
  const newArray = [...arr];
  const index = newArray.findIndex((item) => item.name === itemName);

  if (index !== -1) {
    const item = newArray[index];
    newArray.splice(index, 1);
    newArray.push(item);
  }

  return newArray;
}

export function useGetGoalsConfiguration(
  options?: Omit<
    UseQueryOptions<
      ConnectAdvisorGetGoalTypesResponseDto,
      unknown,
      ConnectAdvisorGetGoalTypesResponseDto['goalTypes'],
      string[]
    >,
    'queryKey' | 'queryFn'
  >
) {
  return useQuery({
    ...getGoalConfigurationsQuery(),
    ...options,
  });
}

export function getGoalsConfigurationLoader(queryClient: QueryClient) {
  return async () => {
    const query = getGoalConfigurationsQuery();
    const data = queryClient.getQueryData(query.queryKey);
    if (data) return data;
    return queryClient.fetchQuery(query);
  };
}

export default useGetGoalsConfiguration;
