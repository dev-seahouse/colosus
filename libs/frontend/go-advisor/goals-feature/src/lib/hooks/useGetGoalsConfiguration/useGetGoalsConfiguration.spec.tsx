import {
  queryClientWrapper,
  renderHook,
  waitFor,
} from '@bambu/react-test-utils';
import useGetGoalsConfiguration from './useGetGoalsConfiguration';

describe('useGetGoalsConfiguration', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetGoalsConfiguration(), {
      wrapper: queryClientWrapper,
    });

    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data).toMatchSnapshot();
  });
});
