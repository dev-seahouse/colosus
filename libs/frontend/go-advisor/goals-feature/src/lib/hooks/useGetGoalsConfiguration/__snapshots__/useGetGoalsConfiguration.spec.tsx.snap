// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetGoalsConfiguration > should render successfully 1`] = `
[
  {
    "description": "Retire comfortably",
    "enabled": true,
    "id": "8d299530-c622-47cb-b425-82429d7443c0",
    "name": "Retirement",
  },
  {
    "description": "Buy a house",
    "enabled": true,
    "id": "d0f191ec-5ddc-4f18-81c6-38b89ac7b5a6",
    "name": "House",
  },
  {
    "description": "Save for college fees",
    "enabled": false,
    "id": "03c46eb2-8004-4fb9-b400-ec3e9903b0db",
    "name": "Education",
  },
  {
    "description": "Just want to grow my wealth",
    "enabled": false,
    "id": "b4dd784c-1743-4161-8b9e-d6e460b6c6c4",
    "name": "Growing Wealth",
  },
  {
    "description": "I have another goal in mind",
    "enabled": false,
    "id": "a4d7abd5-dcbc-44b7-86ca-a0b0cbbd3465",
    "name": "Other",
  },
]
`;
